logserver.c

```diff
@@ -347,127 +372,350 @@ static int logserver_handle_msg(struct logserver *logserver,
+		logserver_handle_msg(msg);
+	}
+out:
+	pv_buffer_drop(buffer);
+}
+
+static void logserver_loop()
+{
+	struct epoll_event ev[LOGSERVER_MAX_EV];
+	int n_events = logserver_epoll_wait(ev);
+
+	if (n_events < 1) {
+		return;
+	}
+
+	int logsock = logserver_g.logsock;
+	int fdsock = logserver_g.fdsock;
+	struct dl_list *tmplst = &logserver_g.tmplst;
+	struct dl_list *fdlst = &logserver_g.fdlst;
+
+	int curfd = -1;
+	for (int i = 0; i < n_events; ++i) {
+		curfd = ev[i].data.fd;
+		if (curfd == logsock || curfd == fdsock) {
+			int fd = logserver_accept_connection(curfd);
+			if (fd < 0) {
+				continue;
+			}
+
+			if (logserver_epoll_add(fd) != 0) {
+				close(fd);
+				continue;
+			}
+
+			if (curfd == fdsock) {
+				struct logserver_fd *lfd =
+					logserver_fd_new(NULL, NULL, fd);
+
+				if (logserver_list_add(tmplst, lfd) != 0) {
+					logserver_fd_free(lfd);
+					logserver_epoll_del(fd);
+					close(fd);
+				}
+			}
+		} else if (logserver_g.out.sfile &&
+			   logserver_list_exists(tmplst, curfd)) {
+			logserver_list_del(tmplst, curfd, NULL);
+			logserver_epoll_del(curfd);
+
+			struct logserver_fd *lfd = logserver_get_fd(curfd);
+
+			if (!lfd) {
+				close(curfd);
+				continue;
+			}
+
+			if (lfd->fd < 0) {
+				pv_log(INFO, "removing fd %s:%s(%d)",
+				       lfd->platform, lfd->src, lfd->fd);
+				logserver_list_del(fdlst, 0, lfd->platform);
+				logserver_fd_free(lfd);
+				close(curfd);
+				continue;
+			}
+
+			if (logserver_list_add(fdlst, lfd) != 0) {
+				pv_log(ERROR, "could not add %s:%s(%d)",
+				       lfd->platform, lfd->src, lfd->fd);
+				logserver_fd_free(lfd);
+				close(curfd);
+				continue;
+			}
 
-	fd = socket(AF_UNIX, SOCK_STREAM, 0);
+			if (logserver_epoll_add(lfd->fd) != 0) {
+				pv_log(ERROR,
+				       "could not add %s:%s(%d), epoll_add error",
+				       lfd->platform, lfd->src, lfd->fd);
+				logserver_fd_free(lfd);
+				close(curfd);
+				continue;
+			}
+
+			pv_log(INFO, "new fd (%d) for %s:%s", lfd->fd,
+			       lfd->platform, lfd->src);
+
+			close(curfd);
+		} else {
+			bool sub = logserver_list_exists(fdlst, curfd);
+			logserver_consume_data(curfd, sub);
+
+			if (!logserver_g.out.sfile || !sub) {
+				logserver_epoll_del(curfd);
+				close(curfd);
+			}
+		}
+	}
+}
+
+static int logserver_open_socket(const char *fname, bool server)
+{
+	struct sockaddr_un addr = { 0 };
+	int fd = socket(AF_UNIX, SOCK_STREAM, 0);
 	if (fd == -1) {
 		pv_log(ERROR, "unable to open control socket: %d", errno);
-		goto out;
+		return -1;
 	}
 
-	unlink(path); // sometimes, the socket file still exists after reboot
-
-	memset(&addr, 0, sizeof(addr));
 	addr.sun_family = AF_UNIX;
-	strncpy(addr.sun_path, path, sizeof(addr.sun_path));
-
-	if (bind(fd, (const struct sockaddr *)&addr, sizeof(addr.sun_path)) ==
-	    -1) {
-		pv_log(ERROR, "unable to bind control socket: %s",
-		       strerror(errno));
-		close(fd);
-		fd = -1;
-		goto out;
-	}
+	// size of sockaddr_un sun_path
+	char path[107];
```

Comment:
better use PATH_MAX here if you are going to use it in pv_paths_pv_file

Reply:
Mmm.. we have a problem here because the `sun_path` member has only 108 bytes, so if I use `PATH_MAX` I could write out of the bounds of the `sun_path` member.

Reply:
Maybe you can apply here the same solution I used here https://github.com/pantavisor/pantavisor/pull/139/files